system/metrics/sinks/log/logSink.cpp

```diff
@@ -47,57 +47,85 @@ void LogMetricSink::doCollection()
 
 void LogMetricSink::writeLogEntry(const std::shared_ptr<IMetric> &pMetric)
 {
-    std::string name = pMetric->queryName();
+    const std::string & name = pMetric->queryName();
+    __uint64 metricValue = pMetric->queryValue();
+    if (ignoreZeroMetrics && (metricValue == 0))
+        return;
+
     const auto & metaData = pMetric->queryMetaData();
+    const std::string * searchKey = &name;
+    //Create a unique id from a combination of the name and the labels so that we can check whether this metric instance has changed since the last time it was logged
+    std::string uid;
+    if (!metaData.empty())
+    {
+        uid.append(name);
+        for (auto &metaDataIt: metaData)
+            uid.append("/").append(metaDataIt.value.c_str());
+        searchKey = &uid;
+    }
+
+    auto match = alreadySeen.find(*searchKey);
+    const auto notFound = alreadySeen.end();
+    //If the values haven't changed then avoid logging an update to the logs
+    if ((match != notFound) && (match->second == metricValue))
+        return;
+
+    StringBuffer output;
+
+    StringBuffer labels;
     for (auto &metaDataIt: metaData)
     {
-        name.append(".").append(metaDataIt.value);
+        if (labels.length() > 0)
+            labels.append(",");
+        labels.appendf("{ \"name\":\"%s\", \"value\": \"%s\" }", metaDataIt.key.c_str(), metaDataIt.value.c_str());
```

Comment:
Treating the meta data like labels is a Prometheus feature. I simply concatenated the meta data to create a unique name that would be easy to grep for in the logs if needed. The framework also concatenates the meta data to create the unique metric name when detecting duplicates during registration. That was another reason the same mechanism was used to create the name for the log entry.

Reply:
Yes.  In this case the format is up to us, and I want it to be able to reflect the most detailed representation - so that the developers can clearly see the logical description of the metric in the logs.